tests/common/connections/ssh_console_conn.py

```diff
@@ -0,0 +1,109 @@
+import time
+import re
+from base_console_conn import BaseConsoleConn
+from netmiko.ssh_exception import NetMikoAuthenticationException
+
+class SSHConsoleConn(BaseConsoleConn):
+    def __init__(self, **kwargs):
+        if not kwargs.has_key("console_server_username") \
+                or not kwargs.has_key("console_server_password"):
+            raise ValueError("Either console_server__username or console_server_password is not set")
+
+        # Console via SSH connection need two groups of user/passwd
+        self.sonic_username = kwargs['username']
+        self.sonic_password = kwargs['password']
+        console_username = kwargs['console_server_username'] + r':p=' + str(kwargs['port'])
+        # Restore the default port for SSH
+        del kwargs['port']
+        kwargs['username'] = console_username
+        kwargs['password'] = kwargs['console_server_password']
+        kwargs['device_type'] = "_ssh"
+        super(SSHConsoleConn, self).__init__(**kwargs)
+
+    def session_preparation(self):
+        self._test_channel_read()
+        self.login_stage_2()
+        self.set_base_prompt()
+        # Clear the read buffer
+        time.sleep(0.3 * self.global_delay_factor)
+        self.clear_buffer()
+
+    def login_stage_2(self,
+                      pri_prompt_terminator=r".*# ",
+                      alt_prompt_terminator=r".*\$ ",
+                      username_pattern=r"(?:user:|username|login|user name)",
+                      pwd_pattern=r"assword",
+                      delay_factor=1,
+                      max_loops=20,
+                      ):
+        """
+        Perform a further login
+        :return: None
+        """
+        """Telnet login. Can be username/password or just password."""
+        delay_factor = self.select_delay_factor(delay_factor)
+        time.sleep(1 * delay_factor)
+
+        output = ""
+        return_msg = ""
+        i = 1
+        while i <= max_loops:
+            try:
+                output = self.read_channel()
+                return_msg += output
+
+                # Search for username pattern / send username
+                if re.search(username_pattern, output, flags=re.I):
```

Comment:
I think we should gracefully handle the case when the line is busy. In that case the username_pattern will not be found.

Reply:
Yes. I agree with you. But there is nothing we can do except waiting for session timeout. There is no output at that case.

Reply:
Oh! In that case, can we do it in two steps? 1 - Login to the console switch 2- Check if the line is busy?

Reply:
I'm not sure if it's possible. For some console server (like mad console), there will be nothing output if the port is busy.  We have no idea whether it's cause by latency or busy line.